clients/src/main/java/org/apache/kafka/common/protocol/types/ArrayOf.java

```diff
@@ -41,6 +41,8 @@ public void write(ByteBuffer buffer, Object o) {
     @Override
     public Object read(ByteBuffer buffer) {
         int size = buffer.getInt();
+        if (size < 0)
+            throw new SchemaException("Array size " + size + " cannot be negative");
```

Comment:
Negative array size is something the protocol wiki doesn't really talk about in detail. The current implementation doesn't allow it but perhaps it should. From a consistency perspective the array specification for bytes and string allow -1 to indicate null. Should arrays in general follow that pattern? This change was made as part of #1095 but may need to be discussed in a KIP unless null was always intended to be -1.